src/sqlfluff/core/parser/grammar/base.py

```diff
@@ -121,30 +134,42 @@ class BaseGrammar(Matchable):
     """
 
     is_meta = False
-    # Are we allowed to refer to keywords as strings instead of only passing
-    # grammars or segments?
-    allow_keyword_string_refs = True
-    equality_kwargs: Tuple[str, ...] = ("optional", "allow_gaps")
+    equality_kwargs: Tuple[str, ...] = ("_elements", "optional", "allow_gaps")
+
+    @overload
+    @staticmethod
+    def _resolve_ref(elem: None) -> None:
+        ...
 
+    @overload
     @staticmethod
-    def _resolve_ref(elem):
+    def _resolve_ref(elem: str) -> "BaseGrammar":
+        ...
+
+    @overload
+    @staticmethod
+    def _resolve_ref(elem: Matchable) -> Matchable:
+        ...
+
+    @overload
```

Comment:
Not sure I understand these overloads? Can you please explain?

Reply:
Yeah maybe they're a bit overkill. I'll remove them. I don't think they're _necessary_ per se.

Reply:
Actually - I tried, they are necessary - but I have managed to reduce them a bit. In essence it's a hint to mypy to say, when you provide `None` you get `None`, otherwise you will always get at least _something_. If I remove them all I get awkward errors.